ENH Array API support for PCA

mtsokol opened the pull request:
#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Based on #25956


#### What does this implement/fix? Explain your changes.

This PR adds PyTorch support (via `array_api_compat`) for PCA. It changes heavy operations (i.e. `svd`) to use proper backend, based on passed array type. Also a unit test is added to assert that PyTorch output is the same as NumPy one.

Solvers support:

|  | NumPy | PyTorch 
| --- | --- | --- 
| full | yes | yes
| randomized | yes | refactor required to adjust `randomized` to array_api
| arpack | yes | no

The `arpack` solver uses `svds` which is not supported by PyTorch (the closest method that I found was `torch.svd_lowrank` that is meant to be used for sparse matrices, but it only computes an approximation). Should I just throw an exception with proper description? (that `arpack` for PyTorch tensors is not supported)

Similar case occurs for `randomized` solver. The `randomized_svd` is using `lu` method for decomposition, whose API differs between SciPy and PyTorch (here, the implementation uses `permute_l=True` parameter that returns `PL, U` instead of `P, L, U`. The PyTorch implementation doesn't support this parameter. Moreover Numpy.linalg does not provide `lu`). Therefore supporting `randomized` would require explicitly checking if PyTorch backend is present.

#### Any other comments?

Please share your feedback!


### TODO

- [x] Improve test coverage for error messages for the unsupported cases
- [x] Manually run the cupy and torch tests on a machine with cuda
- [x] Measure performance impact
   - [x] on CPU with torch
   - [x] on GPU with torch
   - [x] on GPU with cupy
- [x] Rework the tooling to make it possible to test estimator specific methods (e.g. the `get_covariance` / `get_precision` methods)


### Benchmark results


```
Data shape (500000, 1000) and dtype float32 of size 2000.0 MB
PCA(n_components=5, svd_solver='randomized', power_iteration_normalizer='QR')
Fitting PCA(n_components=5) with numpy took 42.591s
Fitting PCA(n_components=5) with numpy and n_threads=1 took 18.953s
Fitting PCA(n_components=5) with numpy and n_threads=4 took 44.034s
Fitting PCA(n_components=5) with torch on CPU took 4.163s
Fitting PCA(n_components=5) with torch on GPU took 0.888s
Fitting PCA(n_components=5) with cupy on GPU took 0.934s

```

=> numpy with MKL has a thread-related performance problem with `float32` data!

EDIT: I tried with OpenBLAS and the numpy code runs in 6 to 8s (not exactly the same machine though). So there is definitely a problem between numpy and MKL on float32 data for this workload.

```
Data shape (500000, 1000) and dtype float64 of size 4000.0 MB
PCA(n_components=5, svd_solver='randomized', power_iteration_normalizer='QR')
Fitting PCA(n_components=5) with numpy took 6.847s
Fitting PCA(n_components=5) with numpy and n_threads=1 took 31.415s
Fitting PCA(n_components=5) with numpy and n_threads=4 took 12.627s
Fitting PCA(n_components=5) with torch on CPU took 4.229s
Fitting PCA(n_components=5) with torch on GPU took 0.912s
Fitting PCA(n_components=5) with cupy on GPU took 0.412s
```

```
Data shape (500000, 1000) and dtype float32 of size 2000.0 MB
PCA(n_components=5, svd_solver='full')
Fitting PCA(n_components=5) with numpy took 24.863s
Fitting PCA(n_components=5) with torch on CPU took 8.832s
Fitting PCA(n_components=5) with torch on GPU took 1.513s
Fitting PCA(n_components=5) with cupy on GPU took 4.109s
Fitting PCA(n_components=5) with cupy with cuML on GPU took 0.683s
```

Environment:

```python
[{'filepath': '/data/parietal/store3/work/ogrisel/mambaforge/envs/py310/lib/libomp.so',
  'internal_api': 'openmp',
  'num_threads': 48,
  'prefix': 'libomp',
  'user_api': 'openmp',
  'version': None},
 {'filepath': '/data/parietal/store3/work/ogrisel/mambaforge/envs/py310/lib/libmkl_rt.so.2',
  'internal_api': 'mkl',
  'num_threads': 48,
  'prefix': 'libmkl_rt',
  'threading_layer': 'gnu',
  'user_api': 'blas',
  'version': '2022.1-Product'}]
```

This machine has a 48 physical core CPU and a NVIDIA A100 GPU.

Benchmark script:

- https://gist.github.com/ogrisel/44bae6f8988abacae358c047d3ecc147

mtsokol committed:
ENH Adds PyTorch support for PCA

mtsokol committed:
ENH Support get_precision and get_covariance

mtsokol commented:
One more question regarding `np.flat` method. In the `get_covariance` method the `np.flat` is used to modify array's diagonal. PyTorch offers `torch.view` instead for that purpose, but Numpy's `view` is completely different (only changing dtype).

I used `xp.reshape`, as it conforms to both libraries to access the diagonal, but it might cause array copying (stated explicitly in docs).

Should I use `np.flat` and `torch.view` and check array type backend explicitly in code? (But it's not desired approach). Or do you know another way for accessing array's diagonal?

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -482,9 +488,12 @@ def _fit(self, X):
             )
+        # Raise an error for torch input and arpack or randomized solver.
+        if self.svd_solver in ["arpack", "randomized"] and _is_torch_namespace(xp):
+            raise TypeError(self._pca_torch_arpack_solver_error_message)
```
For `randomized_svd`, apparently `xp.linalg.lu` is not standardized (yet?):

- https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html

so we might want to implement a namespace specific compatibility stub but we might to defer that to a follow-up PR.

ogrisel replied:
I have opened https://github.com/data-apis/array-api/issues/627 to discuss the possibility to standardize an API for LU factorization.

betatim replied:
> I would rather raise a `ValueError` stating that `svd_solver="arpack"` is only valid when the input data is passed as a numpy array and avoid special casing PyTorch: ideally all Array API compliant namespaces should work, not just PyTorch.

👍  for this. I think having a difference between "works with Numpy" and "works with Array API support enabled" is fine. In the case where there is no Array API support I'd even go as far as making it so that you receive an error if you pass a Numpy array when Array API support is enabled (even though it works if you disable Array API support).

Having Array API support but only for some libraries (say PyTorch but not Cupy) will be confusing and surprising to users.

At the same time I think it is fine to special case things for performance reasons. This means we have one path that works for all Array API compliant inputs and in addition a special PyTorch path that is significantly faster. The important thing is that there is a implementation that works with all Array API compliant inputs/namespaces.

mtsokol replied:
I changed the default normalizer to "qr", as this will allow to use `torch.linalg.rq` and `np.linalg.qr` without much overhead with `xp.linalg.qr`. Also changed `TypeError` to `ValueError`.

ogrisel reviewed (commented):
Thanks for the PR, here is a first batch of feedback.

Out of curiosity did you run some benchmarks both on a CPU and on a GPU?

Also please avoid assuming that `get_namespace` will only be used to add torch support but let's strive to add generic Array API support as much as possible by default. In particular, it would be great to make this PR generic enough to also work with CuPy (and tested accordingly).

> I used xp.reshape, as it conforms to both libraries to access the diagonal, but it might cause array copying (stated explicitly in docs).

> Should I use np.flat and torch.view and check array type backend explicitly in code? (But it's not desired approach). Or do you know another way for accessing array's diagonal?

It's probably acceptable to have namespace specific code paths for performance reasons with a fallback to the default `xp.reshape`. However we should try to factorize those namespace specific hacks in a common private module reused by all scikit-learn estimators that may need  it.

ogrisel commented:
Also @mtsokol be aware that there are related concurrent discussions in other issues that might be of interest for this PR. I tried to label all those issues and PRs with the "Array API" label:

- https://github.com/scikit-learn/scikit-learn/labels/Array%20API

Since Array API support is still quite new and experimental we should try to coordinate between the work of different PRs to make consistent design decisions across estimators.

mtsokol commented:
Hi @ogrisel,

Thank you for your feedback! I think all comments are clear to me, let me follow-up on connected PRs/work to avoid repeating changes, then I will continue this PR. I think a separate module to store all tensor-library specific hacks should help here.

Unfortunately, I don't have access to GPU, but I will perform CPU benchmark comparison across libraries.

ogrisel commented:
About the `np.flat`, maybe we you can implement a private `_flat` helper as done for `_nanmin` in #26243.

thomasjpfan commented on sklearn/decomposition/_base.py:
```diff
@@ -37,13 +37,18 @@ def get_covariance(self):
+            exp_var - self.noise_variance_, xp.zeros_like(exp_var)
+        )
+        cov = (components_.T * exp_var_diff) @ components_
+        # TODO use views instead?
+        cov.reshape(-1)[:: len(cov) + 1] += self.noise_variance_  # modify diag inplace
```
We can not use `cov.reshape` because it is not a part of the Array API spec.

The Array API spec for [reshape](https://data-apis.org/array-api/latest/API_specification/generated/array_api.reshape.html?highlight=reshape) forces a view when `copy=False`, but NumPy's Array API does not follow the spec yet: https://github.com/numpy/numpy/issues/23410.

In the short term, I recommend using a `for` loop over the diagonal when `xp` is not NumPy.

```python
def add_to_diagonal(array, value, xp):
    if _is_numpy_namespace(xp):
        array_np = np.asarray(array)
        array_np.flat[:: array.shape[0] + 1] += value
        return xp.asarray(array_np)
    else:
        for i in range(array.shape[0]):
            array[i, i] += value
```

betatim replied:
Can we put things like this ("code that contains `if numpy: ... else: ...`") into the `_array_api.py` module? Or at least try to place it there to reduce the number of places we have to go hunt for such code fragments?

There can be namespace specific code that we leave in the estimator but I think it is worth making an effort to try and collect as much as possible in one place.

thomasjpfan reviewed (commented):
Thank you for the PR @mtsokol !

I recommend testing with `numpy.array_api` to catch places where the Array API specification is not being followed. Here is an example:

https://github.com/scikit-learn/scikit-learn/blob/fa5f43fa2b060c28b4e71d00837e69d6830c934b/sklearn/tests/test_discriminant_analysis.py#L680-L684

asmeurer commented on sklearn/decomposition/_pca.py:
```diff
@@ -636,12 +649,12 @@ def _fit_truncated(self, X, n_components, svd_solver):
 
         self.explained_variance_ratio_ = self.explained_variance_ / total_var
-        self.singular_values_ = S.copy()  # Store the singular values.
+        self.singular_values_ = xp.asarray(S)  # Store the singular values.
```
If you want to guarantee a copy here you should use `asarray(S, copy=True)`.

mtsokol replied:
Sure! Added.

asmeurer commented on sklearn/decomposition/_base.py:
```diff
@@ -57,26 +62,33 @@ def get_precision(self):
+        precision = components_.T @ xp.linalg.inv(precision) @ components_
         precision /= -(self.noise_variance_**2)
-        precision.flat[:: len(precision) + 1] += 1.0 / self.noise_variance_
+        precision.reshape(-1)[:: len(precision) + 1] += 1.0 / self.noise_variance_
```
reshape is a function in the array API spec (this is the sort of thing you would catch by testing against numpy.array_api).

thomasjpfan replied:
This code is trying to alter the array in place. I think this can work with `xp.reshape(..., copy=False)`, but `copy=False` is not accepted in `numpy.array_api` yet: https://github.com/numpy/numpy/issues/23410.

asmeurer replied:
Oh I see. That needs to be fixed upstream. There are still some things missing from numpy.array_api unfortunately (including all v2022 stuff). Perhaps we should prioritize this for the next NumPy release (when is that again?). CC @rgommers

asmeurer replied:
We could also add a simple `reshape` helper to array_api_compat if the release cadence for fixing numpy.array_api is too slow here.

rgommers replied:
> Perhaps we should prioritize this for the next NumPy release (when is that again?). CC @rgommers

Yes, we should. The release split will happen pretty soon - I haven't seen a release schedule, but I suspect it'll be the second half of May.

ogrisel replied:
+1 but as scikit-learn needs compat with older numpy versions we need the compat workaround suggested in https://github.com/scikit-learn/scikit-learn/pull/26315#discussion_r1186082492 anyway in scikit-learn for the time being.

betatim commented:
I created #26348 to discuss the topic of common tests for Array API suport, hopefully we can converge on something and make a separate PR for this topic.

ogrisel commented:
@mtsokol #26372 was merged in `main`. Feel free to update this PR accordingly: in particular there is a new estimator tag to add to the estimator to have the new common tests run on it.

Also let us know if you want someone else to takeover the PR to address all the comments in the above discussion if you don't have the time to do it yourself.

mtsokol committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented:
For reference, I commented on an older thread about the randomized svd case here: https://github.com/scikit-learn/scikit-learn/pull/26315#discussion_r1232176962

mtsokol committed:
ENH apply review comments

mtsokol commented on sklearn/decomposition/tests/test_pca.py:
```diff
@@ -17,6 +22,52 @@
+@pytest.mark.parametrize("dtype", ["float32", "float64"])
+@pytest.mark.parametrize("svd_solver", PCA_SOLVERS)
+@pytest.mark.parametrize("n_components", range(1, iris.data.shape[1]))
+def test_pca_array_torch(device, dtype, svd_solver, n_components):
```
Sure! I adapted to the new array_api tag, so this test is removed.

mtsokol commented:
> @mtsokol #26372 was merged in `main`. Feel free to update this PR accordingly: in particular there is a new estimator tag to add to the estimator to have the new common tests run on it.
> 
> Also let us know if you want someone else to takeover the PR to address all the comments in the above discussion if you don't have the time to do it yourself.

Hi @ogrisel,
I updated the PR, so `array_api` compatibility is tested in `test_check_estimator_clones` instead of a dedicated `pca` test. Right now it's still the `randomized` solver to be adapted to `array_api` (for example `sklearn.utils.extmath.py` line 808,  uses advanced integer list indexing that is not allowed by `array_api` standard).  

If it's Ok, maybe someone can take this PR from here? I'm Sorry for not communicating that earlier.

ogrisel committed:
Fix multi-fancy indexing via using xp.take on flattened arrays in svd_flip

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -72,6 +72,11 @@ def _is_numpy_namespace(xp):
     return xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}
 
 
+def _is_torch_namespace(xp):
+    """Return True if xp is backed by PyTorch."""
+    return xp.__name__ in {"torch", "array_api_compat.torch"}
+
+
```
This does not seem to be used anymore.

```suggestion
```

ogrisel commented:
I tried to push a fix in the previous commit but apparently it's wrong (it broke many other tests). I will investigate later.

ogrisel committed:
Unit test for svd_flip

ogrisel committed:
Fix stide related logic in call to xp.take with 1d args

ogrisel committed:
Delete dead code

ogrisel commented:
The remaining torch failures happen because by default `n_components=None` which means that we do a non-reducing PCA. The problem is that the last rows `components_` have an explained variance ratio below 1e-16 (as expected) and are therefore highly numerically unstable.

I confirmed that the common test pass with:

```diff
diff --git a/sklearn/utils/estimator_checks.py b/sklearn/utils/estimator_checks.py
index 7d8e673210..bac31d3c1f 100644
--- a/sklearn/utils/estimator_checks.py
+++ b/sklearn/utils/estimator_checks.py
@@ -884,6 +884,10 @@ def check_array_api_input(
     X, y = make_classification(random_state=42)
     X = X.astype(dtype, copy=False)
 
+    if estimator_orig.__class__.__name__ == "PCA":
+        estimator_orig = clone(estimator_orig)
+        estimator_orig.n_components = 2
+
     X = _enforce_estimator_tags_X(estimator_orig, X)
     y = _enforce_estimator_tags_y(estimator_orig, y)
 
@@ -927,7 +931,6 @@ def check_array_api_input(
         "predict_log_proba",
         "predict_proba",
         "transform",
-        "inverse_transform",
     )
 
     for method_name in methods:
```

but I don't want to put estimator specific logic in a common test.

Instead of a common tests, which are typically run with default constructed estimators, we should rather call the `check_array_api_input` checker manually on estimators with specific constructor params where we know the array-api independence is expected to hold numerically.

ogrisel committed:
Run PCA array API tests manually

ogrisel committed:
Do note check concrete values in Array API common test by default

betatim committed:
Move namespace+parameter generation to _array_api

Create a central place to generate all the combinations of (namespace,
device, dtype) that should be tested.

betatim committed:
Rename function

ogrisel committed:
Test .score and .score_samples

ogrisel commented on sklearn/decomposition/_base.py:
```diff
@@ -57,26 +62,36 @@ def get_precision(self):
+        # TODO use views instead?
+        xp.reshape(precision, shape=(-1,), copy=False)[:: len(precision) + 1] += (
+            1.0 / exp_var_diff
+        )
```
This is currently braking the tests. I plan to implement the proposal of
https://github.com/scikit-learn/scikit-learn/pull/26315/files#r1186082492 in the wrapper to fix it.

ogrisel committed:
Progress on fixing .score

ogrisel committed:
Make accuracy_score and score return float explicitly

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel committed:
Fix docstests and test_check_array_api_input

ogrisel committed:
Fix one more doctest fix

ogrisel commented on sklearn/decomposition/tests/test_pca.py:
```diff
@@ -686,3 +689,18 @@ def test_variance_correctness(copy):
+    # generator instead of a for loop
+    estimators = [
+        PCA(n_components=2, svd_solver="full"),
+        # PCA(n_components=2, svd_solver="randomized", power_iteration_normalizer="QR"),
```
This is still not tested and would fail if uncommented. Maybe we could focus this PR on the "full" solver and handle the randomized svd solver in a follow-up PR because this PR is already quite big.

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +386,48 @@ def _expit(X):
+    if sample_weight is not None:
+        return float(sample_score @ sample_weight)
+    else:
+        return float(xp.sum(sample_score))
```
Note: we need to call `float(...)` to convert the output to Python scalar instead of 0-dim Array objects.

The `accuracy_score` docstring states that the return value is of type `float` but is as actually a numpy 0-dim array which might be confusing.

Converting to float is probably saner for user code. Otherwise, some consumer code would need to explicitly move back the scalar code to host memory explicitly to manipulate 0-dim Array instance back from GPU memory if when the input data comes from GPU device memory.

asmeurer replied:
I don't know how this fits into the bigger picture of what scikit-learn is doing, but it's worth noting that calling `float()` would prevent things from working with libraries that use lazy evaluation. There aren't currently any such libraries conforming to the array API, but there likely will be in the future. See the discussion at https://github.com/data-apis/array-api/issues/642.

ogrisel replied:
That's a good point. I will add a note for now. I think we need a dedicated issue in scikit-learn to explore how to use lazy Array API implementations with scikit-learn estimators.

ogrisel replied:
I opened the following issue to track lazy evaluation:

- #26724

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -902,57 +918,107 @@ def check_array_api_input(
-        assert (
-            get_namespace(result_xp)[0] == get_namespace(X_xp)[0]
-        ), f"'{method}' output is in wrong namespace"
+        if method_name in ["score"]:
```
```suggestion
        if method_name == "score":
```

ogrisel committed:
Cosmit

ogrisel committed:
Fix one more doctest fix

ogrisel committed:
Move and update changelog entry

ogrisel committed:
Update the Array API doc page to mention PCA

ogrisel committed:
WIP array api for randomized_svd

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -435,7 +435,7 @@ where :math:`1(x)` is the `indicator function
   0.5
   >>> accuracy_score(y_true, y_pred, normalize=False)
-  2
+  2.0
```
This is impacted because we now call `float(...)` explicitly on the output of `accuracy_score` and `zero_one_loss` to return a Python scalar instead of 0-dim numpy object whose dtype can vary in hard to predict ways depending on the inputs.

ogrisel replied:
Note: the docstring of `accuracy_score` specifies that the return type is float, so this might be considered a bugfix. Not sure if we need a changelog entry for this.

betatim replied:
I guess no one has noticed this difference till now because the 0-dim numpy array is mostly indistinguishable from a float?

ogrisel replied:
Yes, unless it as an int `dtype` as was the case in this particular code snippet prior to this PR :)

betatim commented on sklearn/decomposition/_base.py:
```diff
@@ -57,26 +62,31 @@ def get_precision(self):
 
-        if np.isclose(self.noise_variance_, 0.0, atol=0.0):
-            return linalg.inv(self.get_covariance())
+        if self.noise_variance_ == 0.0:
```
I don't think this is the same as `np.isclose(self.noise_variance_, 0.0, atol=0.0)` because `isclose` has `rtol != 0` by default. However, I am not sure this matters? 

It is a bit annoying that the array API doesn't include a `isclose` :-/

ogrisel replied:
I had not realized that. I think we need our own `_isclose` with the same `atol` and `rtol` logic as numpy's.

betatim commented on sklearn/decomposition/_base.py:
```diff
@@ -147,16 +159,16 @@ def inverse_transform(self, X):
-                    np.sqrt(self.explained_variance_[:, np.newaxis]) * self.components_,
-                )
+                X
+                @ (np.sqrt(self.explained_variance_[:, np.newaxis]) * self.components_)
```
```suggestion
                @ (xp.sqrt(self.explained_variance_[:, xp.newaxis]) * self.components_)
```
no?

thomasjpfan commented on sklearn/decomposition/_pca.py:
```diff
@@ -530,20 +538,20 @@ def _fit_full(self, X, n_components):
         X -= self.mean_
 
-        U, S, Vt = linalg.svd(X, full_matrices=False)
+        U, S, Vt = xp.linalg.svd(X, full_matrices=False)
```
On `main`, this uses `scipy.linalg.svd`. This PR switches to using `numpy.linalg.svd` when container is `ndarray`.

ogrisel replied:
Is this a problem? Both use LAPACK from OpenBLAS in the end nowadays, no?

ogrisel replied:
I special cased the numpy case to continue calling scipy's linalg instead for now.

ogrisel committed:
Merge main

ogrisel committed:
pytest parametrization to run custom Array API checks in test_pca

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel committed:
Pass estimator to check yielder

ogrisel committed:
Update sklearn/decomposition/_base.py

Co-authored-by: Tim Head <betatim@gmail.com>

ogrisel commented:
I am converting this to draft to finalize the randomized svd port in the QR case.

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -188,7 +189,7 @@ def safe_sparse_dot(a, b, *, dense_output=False):
             ret = a_2d @ b
             ret = ret.reshape(*a.shape[:-1], b.shape[1])
         else:
-            ret = np.dot(a, b)
```
This change is probably causing the regression in the coordinate descent solver. I will investigate later.

ogrisel committed:
Fix check_array_api_input and update PCA accordingly

ogrisel committed:
Compare namespace names

ogrisel committed:
Keep on using scipy.linalg.svd in PCA by default

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -773,6 +812,10 @@ def svd_flip(u, v, u_based_decision=True):
     Adjusts the columns of u and the rows of v such that the loadings in the
     columns in u that are largest in absolute value are always positive.
 
+    if u_based_decision is False, then the same sign correction is applied to
```
```suggestion
    If u_based_decision is False, then the same sign correction is applied to
```

ogrisel committed:
cosmetics

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -278,12 +311,12 @@ def randomized_range_finder(
     # Sample the range of A using by linear projection of Q
     # Extract an orthonormal basis
-    Q, _ = linalg.qr(safe_sparse_dot(A, Q), mode="economic")
+    Q, _ = xp.linalg.qr(safe_sparse_dot(A, Q), mode="reduced")
```
/cc @thomasjpfan if you have any opinion.

ogrisel replied:
I changed the code to always keep on using `scipy.linalg` by default.

ogrisel committed:
Test errors when calling PCA with unsupported parameter values

ogrisel committed:
Protect array_api test against missing soft dependency

ogrisel committed:
More consistent use of scipy.linalg when array api is disabled

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel committed:
Improve coverage and simplify code

ogrisel commented:
The tests on the CI should be green and coverage good enough.

However I did not find the time to run the tests or evaluate the performance on GPUs yet.

I updated the description of the PR to reflect what remains to be done. Please feel free to participate in the resolution of the missing items.

ogrisel committed:
Comment about lazy evaluation

ogrisel committed:
Simplify power iteration by using @ instead

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +386,54 @@ def _expit(X):
+
+def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+    # XXX: this function accepts Array API input but returns a Python scalar
+    # float. The coal to float() is convenient because it does hide the need to
```
```suggestion
    # float. The call to float() is convenient because it does hide the need to
```

ogrisel committed:
Update sklearn/utils/_array_api.py

ogrisel committed:
Clean-up left over

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -226,7 +229,7 @@ def randomized_range_finder(
         typically 5 or larger), or 'LU' factorization (numerically stable
         but can lose slightly in accuracy). The 'auto' mode applies no
-        normalization if `n_iter` <= 2 and switches to LU otherwise.
+        normalization if `n_iter` <= 2 and switches to QR otherwise.
```
```suggestion
        normalization if `n_iter` <= 2 and switches to LR otherwise.
```

ogrisel committed:
Fix randomized_range_finder docstring to reflect the latest version of the code

ogrisel committed:
Better not use np.newaxis in Array API code

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel committed:
Preserve device info svd_flip

ogrisel committed:
Update array-api-compat version in build_tools/azure/pylatest_conda_forge_mkl_linux-64_conda.lock

ogrisel committed:
Fix xp.take in sklearn.linear_model by passing axis argument to fit array-api-compat

ogrisel committed:
Fix LinearDiscriminantAnalysis.score to work with cupy

ogrisel commented:
I pushed a few more commits to make all tests pass with PyTorch and cupy on a cuda GPU (tested manually as there is not CI config yet for those).

The benchmark results are good (see the updated description). I discovered a problem with vanilla PCA on `float32` NumPy data on the benchmark machine that has many (48) physical cores. I don't think this is related to this PR. I will need to reproduce on `main` and open a dedicated issue to investigate further. Note: I did not reproduce this problem on my local laptop with 8 cores and OpenBLAS. So it might be a problem only for MKL and/or many core machines.

Note: I also wanted to compare to cuML on the same machine but it does not support `PCA` with `svd_solver="randomized"`. Still it's default solver is as fast as PyTorch or cupy with the randomized solver on the same GPU.

I updated the TODO list items. There are still thinks to improve w.r.t. testing unit functions (instead of estimators) on different Array API implementations.

Furthermore, since I also extended the common test to test the `.score` method of the checked estimators, I had to fix bugs in linear models / helper functions that are not directly related to `PCA`. If reviewers prefer, I could split out those changes in a dedicated PR.

ogrisel committed:
Test get_covariance / get_precision

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on build_tools/azure/pylatest_conda_forge_mkl_linux-64_conda.lock:
```diff
@@ -110,21 +110,22 @@ https://conda.anaconda.org/conda-forge/linux-64/libhwloc-2.9.1-nocuda_h7313eea_6
 https://conda.anaconda.org/conda-forge/linux-64/xcb-util-renderutil-0.3.9-hd590300_1.conda#e995b155d938b6779da6ace6c6b13816
 https://conda.anaconda.org/conda-forge/linux-64/xcb-util-wm-0.4.1-h8ee46fc_1.conda#90108a432fb5c6150ccfee3f03388656
-https://conda.anaconda.org/conda-forge/linux-64/xorg-libx11-1.8.5-h8ee46fc_0.conda#742d9cd4a7da3ac6345f986e5da3b18d
-https://conda.anaconda.org/conda-forge/noarch/array-api-compat-1.2-pyhd8ed1ab_0.conda#3d34f2f6987f8d098ab00198c170a77e
```
I had to update this lock file to get the latest version of `array-api-compat`: I believe this is necessary to get the `axis` parameter for `xp.take`.

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -226,7 +228,7 @@ def randomized_range_finder(
         typically 5 or larger), or 'LU' factorization (numerically stable
         but can lose slightly in accuracy). The 'auto' mode applies no
-        normalization if `n_iter` <= 2 and switches to LU otherwise.
+        normalization if `n_iter` <= 2 and switches to LR otherwise.
```
```suggestion
        normalization if `n_iter` <= 2 and switches to LU otherwise.
```

ogrisel committed:
Typo

ogrisel commented:
Alright, I believe this is ready for review (assuming CI is green). /cc @thomasjpfan @betatim.

betatim commented on sklearn/utils/extmath.py:
```diff
@@ -255,37 +257,71 @@ def randomized_range_finder(
+    # avoid allocating unnecessary memory on the device.
+    if hasattr(A, "device"):
+        if not hasattr(Q, "to_device"):
+            # array_api_compat for PyTorch does not seem to make it possible
+            # to use the intended to_device API:
+            # https://data-apis.org/array-api/2022.12/design_topics/device_support.html
```
The `array_api_compat` [README](https://github.com/data-apis/array-api-compat/tree/main#helper-functions) says that they have a `to_device` helper function to deal with array from libs that do not have `x.to_device` (torch, cupy, numpy). Should we use that instead of `xp.asarray(.., device=A.device)`?

asmeurer replied:
There's also a `device` function to get rid of the other hasattr (I'm curious why you use `hasattr(A, "dtype")`. Is that something we also need a helper for?)

asmeurer replied:
FWIW, this code got me thinking a little bit about the difference between `to_device` and `asarray`. One difference right now is that `asarray` lets you be explicit about copying or not when the array is already on the same device, but I think this is just an underspecification in the standard right now. https://github.com/data-apis/array-api/issues/645

I didn't open an issue for it, but perhaps it would be useful if `to_device` had a dtype argument, similar to `torch.to`, to let you do this whole operation in one step (although I'm honestly surprised that the numpy random generators don't let you natively generate float32).

ogrisel commented on sklearn/decomposition/_base.py:
```diff
@@ -148,16 +160,15 @@ def inverse_transform(self, X):
-                )
-                + self.mean_
+            scaled_components = (
+                xp.sqrt(self.explained_variance_[:, None]) * self.components_
```
I think I observed a problem with `xp.newaxis` (`AttributeError`) for one of the libraries but I don't remember which. I will try again and put a comment if that's still the case.

ogrisel replied:
Oh! I just realized you wrote `np.newaxis` instead of `xp.newaxis`. Indeed you can keep `np.newaxis` since it is actually just a reference to `None` with a more explicit name, although it's a bit confusing to mix `xp` and `np` namespaces in the same snippet. But maybe this is fine here.

ogrisel replied:
I pushed a change to use `np.newaxis` whenever possible.

ogrisel committed:
Add comment to explain why we keep the scipy linalg.svd code path for now

ogrisel committed:
Use array_api_compat.to_device

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

asmeurer commented on sklearn/decomposition/_pca.py:
```diff
@@ -553,16 +567,16 @@ def _fit_full(self, X, n_components):
 
         self.n_samples_ = n_samples
-        self.components_ = components_[:n_components]
+        self.components_ = components_[:n_components, :]
```
Yes, in the array API, indices are only guaranteed to work if all axes are indexed (or if there is an ellipsis).

ogrisel committed:
Improve the to_device helper (with missing docstring)

ogrisel committed:
Use np.newaxis instead of None

ogrisel committed:
Use _is_numpy_namespace / xp.asarray for all numpy backed xp values

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -72,6 +101,13 @@ def _is_numpy_namespace(xp):
 
 
+def _union1d(a, b, xp):
+    if xp.__name__ == "numpy":
```
I agree we can always use `_is_numpy_namespace` and `xp.asarray`.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Tim Head <betatim@gmail.com>

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -528,20 +539,23 @@ def _fit_full(self, X, n_components):
+        if not is_array_api_compliant:
+            U, S, Vt = linalg.svd(X, full_matrices=False)
+        else:
+            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
```
I added a comment on the motivation for keeping this old code-path for now.

ogrisel commented:
Hum, the doc builds now issue new warnings in this PR after merging `main` into this branch, yet the last commit on `main` is green. The warnings seem unrelated with the changes in this PR though. I think they should better be fixed in one or more dedicated PRs.

ogrisel commented:
I re-ran the tests on a cuda machine and `pytest -k array_api -vl sklearn` has all 84 array_api tests pass with the latest state of this PR!

betatim commented:
Now that https://github.com/scikit-learn/scikit-learn/pull/26773 has been turned off again, can you re-trigger the CI?

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -498,12 +502,12 @@ def _should_be_skipped_or_marked(estimator, check):
 
 
-def parametrize_with_checks(estimators):
+def parametrize_with_checks(estimators, check_yielder=None):
```
I will revert this change because it don't need it anymore in the latest version of this PR. So it's probably a YAGNI.

ogrisel committed:
Revert no longer needed change to parametrize_with_checks

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

betatim reviewed (approved):
I contributed a bit of code to this PR, so this review probably doesn't count as a full +1, but still: +1 :D

ogrisel commented:
I think I contributed too much code to vote on this PR anymore but I am morally +1 as well.

Maybe @thomasjpfan would like to have a look?

ogrisel commented:
Otherwise we can wait 1 or 2 weeks and I can re-review myself. I am sure I will be another person by that time given my short attention span :)

thomasjpfan commented on sklearn/decomposition/_base.py:
```diff
@@ -58,26 +63,31 @@ def get_precision(self):
+            components_ = components_ * xp.sqrt(exp_var[:, np.newaxis])
+        exp_var_diff = exp_var - self.noise_variance_
+        exp_var_diff = xp.where(
+            exp_var > self.noise_variance_, exp_var_diff, xp.zeros_like(exp_var)
```
According to the [SPEC](https://data-apis.org/array-api/latest/API_specification/generated/array_api.where.html), `xp.where` uses broadcasting rules, so we do not need to create a whole array of zeros:

```suggestion
            exp_var > self.noise_variance_, exp_var_diff, xp.asarray(0.0)
```

thomasjpfan commented on doc/whats_new/v1.4.rst:
```diff
@@ -43,6 +43,16 @@ Changelog
     where 123456 is the *pull request* number, not the issue number.
 
+
+Code and Documentation Contributors
```
I think this is a merge error. Can we remove this diff?

ogrisel replied:
Done.

thomasjpfan commented on sklearn/utils/extmath.py:
```diff
@@ -255,37 +257,64 @@ def randomized_range_finder(
-        Q = Q.astype(A.dtype, copy=False)
+    # XXX: generate random number directly from xp if it's possible
+    # one day.
+    Q = xp.asarray(random_state.normal(size=(A.shape[1], size)))
+    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
+        # Use float32 computation and components if A has a float32 dtype
+        Q = xp.astype(Q, A.dtype, copy=False)
+
+    # Move Q to device if needed only after converting to float32 if needed to
+    # avoid allocating unnecessary memory on the device.
+    Q = to_device(Q, getattr(A, "device", None))
```
I think this whole section can be simplified to:

```python
    # XXX: generate random number directly from xp if it's possible
    # one day.
    Q_dtype = None
    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
        # Use float32 computation and components if A has a float32 dtype
        Q_dtype = A.dtype

    Q = xp.asarray(
        random_state.normal(size=(A.shape[1], size)), dtype=Q_dtype, device=device(A)
    )
```

where `device` is from `sklearn/utils/_array_api.py`. `device` would need to be updated to:

```python
def device(x):
    if isinstance(x, (numpy.ndarray, numpy.generic)) or issparse(x):
        return "cpu"
    return x.device
```

This also means we do not need a `to_device` function anymore.

asmeurer replied:
It's not clear that that's the case, because asarray() in the spec says that it might not support downcasts. See the discussion at https://github.com/data-apis/array-api/issues/647 (I believe this may be discussed at the next arrays consortium meeting).

ogrisel replied:
Thanks for your input @asmeurer. I added a note in 4093a6c for now.

thomasjpfan commented on sklearn/decomposition/_base.py:
```diff
@@ -58,26 +63,31 @@ def get_precision(self):
-        if np.isclose(self.noise_variance_, 0.0, atol=0.0):
-            return linalg.inv(self.get_covariance())
+        if self.noise_variance_ == 0.0:
+            return xp.linalg.inv(self.get_covariance())
```
On `main`, this uses `scipy.linalg.inv` this PR using `np.linalg.inv`. Should we condition on `is_array_api_compliant` here to use `scipy.linalg.inv`?

ogrisel replied:
Done. I also updated the subsequent `xp.linalg.inv`.

thomasjpfan reviewed (commented):
Overall this PR looks good.

ogrisel committed:
Remove changelog merge typo

ogrisel committed:
Leverage broadcasting to sparse a temp allocation

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel committed:
Keep on using scipy.linalg.inv in PCA.get_covariance by default

ogrisel committed:
Add not on combined dtype conversion and device move

ogrisel commented on sklearn/decomposition/_base.py:
```diff
@@ -38,13 +39,18 @@ def get_covariance(self):
+            components_ = components_ * xp.sqrt(exp_var[:, np.newaxis])
+        exp_var_diff = exp_var - self.noise_variance_
+        exp_var_diff = xp.where(
+            exp_var > self.noise_variance_, exp_var_diff, xp.zeros_like(exp_var)
```
```suggestion
            exp_var > self.noise_variance_, exp_var_diff, xp.asarray(0.0)
```

ogrisel committed:
Spare one more temporary allocation.

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

thomasjpfan commented on sklearn/utils/extmath.py:
```diff
@@ -255,37 +257,70 @@ def randomized_range_finder(
+    # from float64 to float32 in asarray might not always be accepted as only
+    # casts following type promotion rules are guarateed to work.
+    # https://github.com/data-apis/array-api/issues/647
+    Q = to_device(Q, getattr(A, "device", None))
```
I am concerned with `A.device` existing and `get_config()["array_api_dispatch"]` is False. If `array_api_compat` is not installed,  then `to_device` can fail.

As a workaround, I think we need to condition on `is_array_api_compliant` here or in `to_device`.

ogrisel replied:
Should be addressed. I re-ran the array_api test on a cuda machine to make sure that this still works as expected with GPU-backed libraries.

ogrisel committed:
Assume to_device is always called under an is_array_api_compliant condition

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -472,13 +508,19 @@ def randomized_svd(
     # compute the SVD on the thin matrix: (k + p) wide
-    Uhat, s, Vt = linalg.svd(B, full_matrices=False, lapack_driver=svd_lapack_driver)
+    xp, is_array_api_compliant = get_namespace(B)
+    if _is_numpy_namespace(xp) and not is_array_api_compliant:
+        Uhat, s, Vt = linalg.svd(
+            B, full_matrices=False, lapack_driver=svd_lapack_driver
+        )
+    else:
+        Uhat, s, Vt = xp.linalg.svd(B, full_matrices=False)
```
@thomasjpfan I have the feeling that `_is_numpy_namespace(xp)` is redundant here. Do you envision a future where `is_array_api_compliant is False` and `xp.__namespace__ != "numpy"`?

Otherwise:


```suggestion
    if is_array_api_compliant:
        Uhat, s, Vt = xp.linalg.svd(B, full_matrices=False)
    else:
        # Rely on scipy.linalg instead of numpy.linalg to avoid introducing
        # a behavior change with previous versions of scikit-learn when
        # array_api_dispatch is False.
        Uhat, s, Vt = linalg.svd(
            B, full_matrices=False, lapack_driver=svd_lapack_driver
        )
```

betatim replied:
We should add a comment for people from the future. The way I understand it is that if the input is Numpy or Array API support is turned off we want to use the `svd` from scipy. So that the results don't change compared to what is in scikit-learn now.

In the other cases (not a Numpy array and Array API support turned on) we want to use the `svd` from `xp.linalg`.

But I think neither of the proposed bits of code make that obvious. They all check something that is highly correlated with this, but not exactly this. Hence leaving a comment for people from the future.

ogrisel replied:
> if the input is Numpy or Array API support is turned off we want to use the svd from scipy.

Not exactly: at the moment, if `xp.__namespace__ == "numpy"` (but not `xp.__namespace__ == "numpy.array_api"`) and Array API dispatch is turned off, then we use scipy. Otherwise we use `xp.linalg` (e.g. `numpy.array_api.linalg` when `xp.__namespace__ == "numpy.array_api"`).

ogrisel replied:
BTW: I added the comment in the last commit: e30bfa8 (without changing the condition).

ogrisel committed:
Extend the common tests to handle the case when array api is used with raw numpy arrays without device attribute

ogrisel committed:
Extend the common tests to handle the case when array api is used with raw numpy arrays without device attribute

ogrisel commented:
@thomasjpfan actually I still need to protect against missing `device` attributes even when `is_array_api_compliant` is true because the of the `xp.__namespace__ == "numpy"` case (with the numpy wrapper.

I realized that we did not have tests for this case. So I enabled this case in e30bfa8 (and updated the code accordingly). I now have all 92 tests pass (for `pytest -k array_api sklearn`) on a cuda machine with all the soft dependencies.

ogrisel commented:
Note that `array_api_compat` provides a `device` helper function that returns `"cpu"` for numpy arrays and `array.device` otherwise. We could do something similar instead of using `getattr(array, "device", None)` if you prefer.

ogrisel committed:
Simplify condition to protect scipy.linalg.svd in randomized_svd

ogrisel committed:
Merge branch 'main' into feature/array_api_compat_pca

betatim commented:
> Note that `array_api_compat` provides a `device` helper function that returns `"cpu"` for numpy arrays and `array.device` otherwise. We could do something similar instead of using `getattr(array, "device", None)` if you prefer.

Can't we use `device` from `array_api_compat`?

ogrisel commented:
> Can't we use device from array_api_compat?

We would cannot import from `array_api_compat` at the top of the file (because we need to protect the import behind `sklearn.get_config("array_api_dispatch")` or equivalently `is_array_api_compliant`). Hence it would be too ugly/verbose to use `array_api_compat.device` with a protected lazy import each time.

thomasjpfan commented on sklearn/utils/extmath.py:
```diff
@@ -255,37 +257,71 @@ def randomized_range_finder(
+    # from float64 to float32 in asarray might not always be accepted as only
+    # casts following type promotion rules are guarateed to work.
+    # https://github.com/data-apis/array-api/issues/647
+    if is_array_api_compliant:
+        Q = to_device(Q, getattr(A, "device", None))
```
At this point `Q` has the correct dtype, can we use `asarray` to do the device transfer?

```suggestion
    Q = xp.asarray(Q, device=device(A))
```

> Note that array_api_compat provides a device helper function that returns "cpu" for numpy arrays and array.device otherwise. We could do something similar instead of using getattr(array, "device", None) if you prefer.

We already have it in `device` defined in `sklearn.utils._array_api`.

ogrisel replied:
Alright, let's do this then.

ogrisel replied:
Done in b881ff1. This is indeed simpler this way.

thomasjpfan commented on sklearn/utils/extmath.py:
```diff
@@ -255,37 +257,71 @@ def randomized_range_finder(
+    # from float64 to float32 in asarray might not always be accepted as only
+    # casts following type promotion rules are guarateed to work.
+    # https://github.com/data-apis/array-api/issues/647
+    if is_array_api_compliant:
+        Q = to_device(Q, getattr(A, "device", None))
```
At this point `Q` has the correct dtype, can we use `asarray` to do the device transfer?

```suggestion
    Q = xp.asarray(Q, device=device(A))
```

> Note that array_api_compat provides a device helper function that returns "cpu" for numpy arrays and array.device otherwise. We could do something similar instead of using getattr(array, "device", None) if you prefer.

We already have it in `device` defined in `sklearn.utils._array_api`.

thomasjpfan reviewed (commented):
I'm trying to push for us to not need `to_device` and use `xp.asarray` for device transfer. The rest of the PR looks good to go.

I ran the `array_api` test locally with all the dependencies and everything passes.

ogrisel committed:
Use xp.asarray + device instead of a new to_device helper

ogrisel committed:
Fix randomized_range_finder with sparse matrices

thomasjpfan reviewed (approved):
LGTM, thank you everyone for working on this!

betatim commented:
Whoop whoop! Nice work everyone on yet another PR that turns out to involve a lot more work than you originally thought :D
